### PR TITLE
Governance factory

### LINKED ISSUE #280: Governance factory
All governance related contracts should be deployed by a factory that will be called by the old governance.

**Acceptance Criteria (TBD)**
- [ ] Deploys Emission contract
- [ ] Deploys Airgrab
- [ ] Deploys Treasury
- [ ] Deploys a proxy for Vesting contracts
- [ ] Deploys Mento Token
- [ ] Deploys Governor
- [ ] Deploys Locking contracts
- [ ] Initiates upgradeable contracts
- [ ] Tests

### PR DESCRIPTION
**Pull request by baroooo**
### Description

A factory contract that deploys and configures all token and governance related contracts.
It will be initiated by Celo governance. 
Treasury and vesting contracts won't be created in this PR. Those changes will be added later.

### Other changes

Airgrab's init function is updated to resolve triple circular dependence between token, locking and airgrab
Governor config is parameterized to keep all config in the Factory
Added TODO comments as reminders
The branch is based on locking-contracts branch so locking related changes can be ignored in this review (contracts/governance/locking, test/Locking, test/mocks, test/utils)

### Tested

Tested with unit tests
The issue #275 aims to add integration tests that will test the main flows

### Related issues

- Fixes #280

### CODE REVIEW

File: test/utils/TestLocking.sol [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  function getLockTest(
+    uint96 amount,
+    uint32 slope,
+    uint32 cliff
+  ) external view returns (uint96 lockAmount, uint96 lockSlope) {
+    (lockAmount, lockSlope) = getLock(amount, slope, cliff);
+  }
+
+  function getEpochShift() internal view override returns (uint32) {
+    return epocShift;
```
Comment (nvtaveras) [inline test/utils/TestLocking.sol:27]: nit: epocShift?

**Review by nvtaveras — COMMENTED** [review]
Left a few small comments but overall looks good. Would be nice to have a 2nd pass once the locking branch is merged

File: contracts/governance/Factory.sol [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+  Locking public locking;
+
+  bool public initialized; // Indicates if the governance system has been created
+  address public vesting;
+  address public treasury;
+
+  // Airgrab configuration
+  uint32 public constant AIRGRAB_LOCK_SLOPE = 104; // Slope duration for the airgrabed tokens in weeks
+  uint32 public constant AIRGRAB_LOCK_CLIFF = 0; // Cliff duration for the airgrabed tokens
+  uint256 public constant AIRGRAB_DURATION = 365 days; // Duration for the airgrab
```
[the remark this answers is not in the document]
Comment (baroooo) [reply contracts/governance/Factory.sol:49]: yes, after one year remaining funds will be drained

File: contracts/governance/Factory.sol [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+  uint256 public constant FRACTAL_MAX_AGE = 180 days; // Maximum age of the kyc for the airgrab
+
+  // Timelock configuration
+  uint256 public constant TIMELOCK_DELAY = 2 days; // Delay duration for the timelock
+
+  // Governor configuration
+  uint256 public constant GOVERNOR_VOTING_DELAY = 1; // Voting start the next block
+  uint256 public constant GOVERNOR_VOTING_PERIOD = 120_960; // Voting period for the governor (7 days in blocks CELO)
+  uint256 public constant GOVERNOR_PROPOSAL_THRESHOLD = 1_000e18; // Proposal threshold for the governor
```
[the remark this answers is not in the document]
Comment (baroooo) [reply contracts/governance/Factory.sol:58]: it is tentatively decided and noted here:
https://www.notion.so/Mento-Token-and-Governance-ae6acced7b4d465db90f92f1408ee23c?pvs=4#ac6b141b682b4c1b9986c2d50dcca811

**Review by philbow61 — APPROVED** [review]

File: test/governance/Factory.t.sol [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+  function test_createGovernance_whenCalledTwice_shouldRevert() public i_setUp {
+    vm.prank(owner);
+    cg_subject();
+
+    vm.prank(owner);
+    vm.expectRevert("Factory: governance already created");
+    cg_subject();
+  }
+}
```
[the remark this answers is not in the document]
Comment (philbow61) [reply test/governance/Factory.t.sol:94]: * timelock

File: contracts/governance/Factory.sol [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+      AIRGRAB_LOCK_SLOPE
+    );
+    mentoToken = new MentoToken(vesting, address(airgrab), treasury, address(emission));
+    timelockController = new TimelockController();
+    mentoGovernor = new MentoGovernor();
+    locking = new Locking();
+
+    // Initializations
+    airgrab.initialize(address(mentoToken), address(locking), treasury);
+    emission.setTokenContract(address(mentoToken));
```
[the remark this answers is not in the document]
Comment (baroooo) [reply contracts/governance/Factory.sol:108]: yes it is an option, why I keep them separate is token contract will be set in init but emission target will be also set in future independent of the token.

File: contracts/governance/locking/LockingBase.sol [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+    uint32 shifted = ts - (getEpochShift());
+    return shifted / WEEK - uint32(startingPointWeek);
+  }
+
+  /**
+   * @notice method returns the amount of blocks to shift locking epoch to.
+   * we move it to 00-00 UTC Friday by shifting 3564 blocks (approx) (CELO)
+   */
+  function getEpochShift() internal view virtual returns (uint32) {
+    return 3564;
```
[the remark this answers is not in the document]
Comment (baroooo) [reply contracts/governance/locking/LockingBase.sol]: we have an issue to update it

**Pull request merged**